src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReflectTypeDescriptionProvider.cs

```diff
@@ -93,46 +94,84 @@ internal ReflectTypeDescriptionProvider()
 
         private static Hashtable EditorTables => LazyInitializer.EnsureInitialized(ref s_editorTables, () => new Hashtable(4));
 
+        /// <summary>
+        /// Provides a way to create <see cref="TypeConverter"/> instances, and cache them where applicable.
+        /// </summary>
+        private class IntrinsicTypeConverterData
+        {
+            private readonly Func<Type, TypeConverter> _constructionFunc;
+
+            private readonly bool _cacheConverterInstance;
+
+            private TypeConverter _converterInstance;
+
+            /// <summary>
+            /// Creates a new instance of <see cref="IntrinsicTypeConverterData"/>.
+            /// </summary>
+            /// <param name="constructionFunc">
+            /// A func that creates a new <see cref="TypeConverter"/> instance.
+            /// </param>
+            /// <param name="cacheConverterInstance">
+            /// Indicates whether to cache created <see cref="TypeConverter"/> instances. This is false when the converter handles multiple types,
+            /// specifically <see cref="EnumConverter"/>, <see cref="NullableConverter"/>, and <see cref="ReferenceConverter"/>.
+            /// </param>
+            public IntrinsicTypeConverterData(Func<Type, TypeConverter> constructionFunc, bool cacheConverterInstance = true)
+            {
+                _constructionFunc = constructionFunc;
+                _cacheConverterInstance = cacheConverterInstance;
+            }
+
+            public TypeConverter GetOrCreateConverterInstance(Type genericParameterType)
+            {
+                if (_converterInstance == null || !_cacheConverterInstance)
+                {
+                    _converterInstance = _constructionFunc(genericParameterType);
```

Comment:
You shouldn't be caching `_converterInstance` when `_cacheConverterInstance is false`.